Error: Anomaly "Uncaught exception Not_found." on `Declare Module`

<!-- Thank you for reporting a bug to Coq! -->

#### Description of the problem

I'm guessing this is a race between a change I made to fiat-crypto in https://github.com/mit-plv/fiat-crypto/pull/1508 3 days ago, and #16794 which passed the bench 4 days ago but was merged more recently.  The line https://github.com/mit-plv/fiat-crypto/blob/master/src/Assembly/Symbolic.v#L1128 (a `Declare Module`) fails with `Error: Anomaly "Uncaught exception Not_found."`

Log [here](https://github.com/mit-plv/fiat-crypto/actions/runs/3489978378/jobs/5840799203)

cc @SkySkimmer 

#### Coq Version

e043c3e206f1f64303508e89b3501aade7f58a06 (a7ec416deffe000008ce5006e4b75a5a118be0ca works fine)

Minimization finally succeeded!
        @JasonGross, Minimized File /home/coq/.opam/4.13.1+flambda/.opam-switch/build/coq-fiat-crypto.dev/src/Assembly/Symbolic.v (full log [on GitHub Actions](https://github.com/coq-community/run-coq-bug-minimizer/actions/runs/3501712696))

<details><summary>Minimized Coq File (consider adding this file to the test-suite)</summary>

```coq
(* -*- mode: coq; coq-prog-args: ("-emacs" "-q" "-w" "+implicit-core-hint-db,+implicits-in-term,+non-reversible-notation,+deprecated-intros-until-0,+deprecated-focus,+unused-intro-pattern,+variable-collision,+unexpected-implicit-declaration,+omega-is-deprecated,+deprecated-instantiate-syntax,+non-recursive,+undeclared-scope,+deprecated-hint-rewrite-without-locality,+deprecated-hint-without-locality,+deprecated-instance-without-locality,+deprecated-typeclasses-transparency-without-locality,unsupported-attributes" "-w" "-notation-overridden,-deprecated-hint-constr,-fragile-hint-constr,-native-compiler-disabled,-ambiguous-paths,-masking-absolute-name" "-w" "-deprecated-native-compiler-option" "-R" "/home/coq/.opam/4.13.1+flambda/.opam-switch/build/coq-fiat-crypto.dev/src" "Crypto" "-Q" "/github/workspace/cwd" "Top" "-Q" "/home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/Bignums" "Bignums" "-Q" "/home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/Coqprime" "Coqprime" "-Q" "/home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/Ltac2" "Ltac2" "-Q" "/home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/Rewriter" "Rewriter" "-Q" "/home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/coqutil" "coqutil" "-top" "Crypto.Assembly.Symbolic" "-native-compiler" "ondemand") -*- *)
(* File reduced by coq-bug-minimizer from original input, then from 4434 lines to 12 lines, then from 17 lines to 11 lines *)
(* coqc version 8.17+alpha compiled with OCaml 4.13.1
   coqtop version 5fd048cce6a4:/home/coq/.opam/4.13.1+flambda/.opam-switch/build/coq-core.dev/_build/default,master (be037486c27f29dc0bcfbfd742c0303af295e23f)
   Expected coqc runtime on this file: 0.162 sec *)
Module Type DagDefinitions.
End DagDefinitions.
Module DagExtraDefinitions (Import dag : DagDefinitions).

  Module Type EagerDagDefinitions.
  End EagerDagDefinitions.
End DagExtraDefinitions.
  Include DagExtraDefinitions.
  Declare Module eager : EagerDagDefinitions.
```
</details>

<details><summary>Intermediate Coq File (useful for debugging if minimization did not go as far as you wanted)</summary>

```coq

```
</details>

<details><summary>Build Log (contains the Coq error message)</summary>

```
+ [33;1m(/github/workspace/run-script.sh @ line 51) $[0m ocamlc -config
version: 4.13.1
standard_library_default: /home/coq/.opam/4.13.1+flambda/lib/ocaml
standard_library: /home/coq/.opam/4.13.1+flambda/lib/ocaml
ccomp_type: cc
c_compiler: gcc
ocamlc_cflags: -O2 -fno-strict-aliasing -fwrapv -pthread -fPIC 
ocamlc_cppflags: -D_FILE_OFFSET_BITS=64 
ocamlopt_cflags: -O2 -fno-strict-aliasing -fwrapv -pthread -fPIC 
ocamlopt_cppflags: -D_FILE_OFFSET_BITS=64 
bytecomp_c_compiler: gcc -O2 -fno-strict-aliasing -fwrapv -pthread -fPIC  -D_FILE_OFFSET_BITS=64 
native_c_compiler: gcc -O2 -fno-strict-aliasing -fwrapv -pthread -fPIC  -D_FILE_OFFSET_BITS=64 
bytecomp_c_libraries: -lm -ldl  -lpthread
native_c_libraries: -lm -ldl 
native_pack_linker: ld -r -o 
ranlib: ranlib
architecture: amd64
model: default
int_size: 63
word_size: 64
system: linux
asm: as
asm_cfi_supported: true
with_frame_pointers: false
ext_exe: 
ext_obj: .o
ext_asm: .s
ext_lib: .a
ext_dll: .so
os_type: Unix
default_executable_name: a.out
systhread_supported: true
host: x86_64-pc-linux-gnu
target: x86_64-pc-linux-gnu
flambda: true
safe_string: true
default_safe_string: true
flat_float_array: true
function_sections: true
afl_instrument: false
windows_unicode: false
supports_shared_libraries: true
exec_magic_number: Caml1999X030
cmi_magic_number: Caml1999I030
cmo_magic_number: Caml1999O030
cma_magic_number: Caml1999A030
cmx_magic_number: Caml1999y030
cmxa_magic_number: Caml1999z030
ast_impl_magic_number: Caml1999M030
ast_intf_magic_number: Caml1999N030
cmxs_magic_number: Caml1999D030
cmt_magic_number: Caml1999T030
linear_magic_number: Caml1999L030
+ [33;1m(/github/workspace/run-script.sh @ line 52) $[0m coqc --config
MINIMIZER_DEBUG_EXTRA: coqc: /home/coq/.opam/4.13.1+flambda/bin/coqc
MINIMIZER_DEBUG_EXTRA: coqpath: 
MINIMIZER_DEBUG_EXTRA: pwd: PWD=/github/workspace
MINIMIZER_DEBUG_EXTRA: exec: /home/coq/.opam/4.13.1+flambda/bin/coqc.orig --config 
MINIMIZER_DEBUG: info: /tmp/tmp-coqbot-minimizer.lwYBH9Qiez
COQLIB=/home/coq/.opam/4.13.1+flambda/lib/coq/
COQCORELIB=/home/coq/.opam/4.13.1+flambda/lib/coq/../coq-core/
DOCDIR=/home/coq/.opam/4.13.1+flambda/share/doc/
OCAMLFIND=/home/coq/.opam/4.13.1+flambda/bin/ocamlfind
CAMLFLAGS=-thread -rectypes -bin-annot -strict-sequence -w -a+1..3-4+5..8-9+10..26-27+28..39-40-41-42+43-44-45+46..47-48+49..57-58+59..66-67-68+69-70
WARN=-warn-error +a-3
HASNATDYNLINK=true
COQ_SRC_SUBDIRS=boot config lib clib kernel library engine pretyping interp gramlib parsing proofs tactics toplevel printing ide stm vernac plugins/btauto plugins/cc plugins/derive plugins/extraction plugins/firstorder plugins/funind plugins/ltac plugins/ltac2 plugins/micromega plugins/nsatz plugins/ring plugins/rtauto plugins/ssr plugins/ssrmatching plugins/syntax
COQ_NATIVE_COMPILER_DEFAULT=no
+ [33;1m(/github/workspace/run-script.sh @ line 53) $[0m coqc --version
MINIMIZER_DEBUG_EXTRA: coqc: /home/coq/.opam/4.13.1+flambda/bin/coqc
MINIMIZER_DEBUG_EXTRA: coqpath: 
MINIMIZER_DEBUG_EXTRA: pwd: PWD=/github/workspace
MINIMIZER_DEBUG_EXTRA: exec: /home/coq/.opam/4.13.1+flambda/bin/coqc.orig --version 
MINIMIZER_DEBUG: info: /tmp/tmp-coqbot-minimizer.Z2BWysmctH
The Coq Proof Assistant, version 8.17+alpha
compiled with OCaml 4.13.1
+ [33;1m(/github/workspace/run-script.sh @ line 54) $[0m coqtop
+ [33;1m(/github/workspace/run-script.sh @ line 54) $[0m true
MINIMIZER_DEBUG_EXTRA: coqc: /home/coq/.opam/4.13.1+flambda/bin/coqtop
MINIMIZER_DEBUG_EXTRA: coqpath: 
MINIMIZER_DEBUG_EXTRA: pwd: PWD=/github/workspace
MINIMIZER_DEBUG_EXTRA: exec: /home/coq/.opam/4.13.1+flambda/bin/coqtop.orig 
MINIMIZER_DEBUG: info: /tmp/tmp-coqbot-minimizer.i8HySK1jE2
Welcome to Coq 5fd048cce6a4:/home/coq/.opam/4.13.1+flambda/.opam-switch/build/coq-core.dev/_build/default,master (be037486c27f29dc0bcfbfd742c0303af295e23f)

Coq < 
+ [33;1m(/github/workspace/run-script.sh @ line 56) $[0m source /github/workspace/coqbot.sh
++ [33;1m(/github/workspace/run-script.sh @ line 1) $[0m opam install -y coq-fiat-crypto.dev
The following actions will be performed:
  - install coq-coqprime    dev [required by coq-fiat-crypto]
  - install coq-rewriter    dev [required by coq-fiat-crypto]
  - install conf-findutils  1   [required by coq-fiat-crypto]
  - install coq-coqutil     dev [required by coq-fiat-crypto]
  - install coq-fiat-crypto dev
===== 5 to install =====

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
-> installed conf-findutils.1
-> retrieved coq-coqutil.dev  (git+https://github.com/mit-plv/coqutil.git#tested)
-> retrieved coq-coqprime.dev  (git+https://github.com/thery/coqprime.git#master)
-> retrieved coq-rewriter.dev  (git+https://github.com/mit-plv/rewriter.git#master)
-> retrieved coq-fiat-crypto.dev  (git+https://github.com/mit-plv/fiat-crypto.git#master)
-> installed coq-coqutil.dev
-> installed coq-coqprime.dev
-> installed coq-rewriter.dev
[ERROR] The compilation of coq-fiat-crypto.dev failed at "make -j1 EXTERNAL_DEPENDENCIES=1 SKIP_BEDROCK2=1 coq-without-bedrock2 standalone-ocaml".

#=== ERROR while compiling coq-fiat-crypto.dev ================================#
context              2.1.3 | linux/x86_64 | ocaml-option-flambda.1 ocaml-variants.4.13.1+options | https://coq.inria.fr/opam/extra-dev#2022-11-15 19:00
path                 ~/.opam/4.13.1+flambda/.opam-switch/build/coq-fiat-crypto.dev
command              /usr/bin/make -j1 EXTERNAL_DEPENDENCIES=1 SKIP_BEDROCK2=1 coq-without-bedrock2 standalone-ocaml
exit-code            2
env-file             ~/.opam/log/coq-fiat-crypto-848-c502fd.env
output-file          ~/.opam/log/coq-fiat-crypto-848-c502fd.out
### output ###
[...]
MINIMIZER_DEBUG: info: /tmp/tmp-coqbot-minimizer.Wi40383Kkk
COQC src/Assembly/Symbolic.v
MINIMIZER_DEBUG_EXTRA: coqc: /home/coq/.opam/4.13.1+flambda/bin/coqc
MINIMIZER_DEBUG_EXTRA: coqpath: 
MINIMIZER_DEBUG_EXTRA: pwd: PWD=/home/coq/.opam/4.13.1+flambda/.opam-switch/build/coq-fiat-crypto.dev
MINIMIZER_DEBUG_EXTRA: exec: /home/coq/.opam/4.13.1+flambda/bin/coqc.orig -q -w +implicit-core-hint-db\,+implicits-in-term\,+non-reversible-notation\,+deprecated-intros-until-0\,+deprecated-focus\,+unused-intro-pattern\,+variable-collision\,+unexpected-implicit-declaration\,+omega-is-deprecated\,+deprecated-instantiate-syntax\,+non-recursive\,+undeclared-scope\,+deprecated-hint-rewrite-withou[...]
MINIMIZER_DEBUG: info: /tmp/tmp-coqbot-minimizer.etH5ew7BWu
File "./src/Assembly/Symbolic.v", line 1445, characters 2-45:
Error: Anomaly "Uncaught exception Not_found."
Please report at http://coq.inria.fr/bugs/.

make: *** [Makefile.coq:793: src/Assembly/Symbolic.vo] Error 129



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+- The following actions failed
| - build coq-fiat-crypto dev
+- 
+- The following changes have been performed
| - install conf-findutils 1
| - install coq-coqprime   dev
| - install coq-coqutil    dev
| - install coq-rewriter   dev
+- 

The former state can be restored with:
    /usr/local/bin/opam.orig switch import "/home/coq/.opam/4.13.1+flambda/.opam-switch/backup/state-20221119025217.export"
```
</details>

<details><summary>Minimization Log (truncated to last 8.0KiB; full 162KiB file on <a href="https://github.com/coq-community/run-coq-bug-minimizer/actions/runs/3501712696">GitHub Actions Artifacts</a> under <code>bug.log</code>)</summary>

```
q KEEP_ERROR=1 bug_01.glob
getting bug_01.glob (/github/workspace/cwd/bug_01.glob)
getting /github/workspace/cwd/bug_01.glob
[92m
Sanity check passed.[0m

Now, I will attempt to strip repeated newlines and trailing spaces from this file...

No strippable newlines or spaces.

Now, I will attempt to strip the comments from this file...
[92m
Succeeded in stripping comments.[0m

Now, I will attempt to factor out all of the [Require]s...
getting bug_01.v (/github/workspace/cwd/bug_01.v)
NOTE: The file bug_01.v is very new (1668833524, 0 seconds old), delaying until it's a bit older
/home/coq/.opam/4.13.1+flambda/bin/coq_makefile COQC = /home/coq/.opam/4.13.1+flambda/bin/coqc.orig -o Makefilefg49ymn_.coq -R /home/coq/.opam/4.13.1+flambda/.opam-switch/build/coq-fiat-crypto.dev/src Crypto -Q /github/workspace/cwd Top -Q /home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/Bignums Bignums -Q /home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/Coqprime Coqprime -Q /home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/Ltac2 Ltac2 -Q /home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/Rewriter Rewriter -Q /home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/coqutil coqutil -R /home/coq/.opam/4.13.1+flambda/.opam-switch/build/coq-fiat-crypto.dev/src Crypto -Q /github/workspace/cwd Top -Q /home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/Bignums Bignums -Q /home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/Coqprime Coqprime -Q /home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/Ltac2 Ltac2 -Q /home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/Rewriter Rewriter -Q /home/coq/.opam/4.13.1+flambda/lib/coq/user-contrib/coqutil coqutil -arg -q -arg -w -arg +implicit-core-hint-db,+implicits-in-term,+non-reversible-notation,+deprecated-intros-until-0,+deprecated-focus,+unused-intro-pattern,+variable-collision,+unexpected-implicit-declaration,+omega-is-deprecated,+deprecated-instantiate-syntax,+non-recursive,+undeclared-scope,+deprecated-hint-rewrite-without-locality,+deprecated-hint-without-locality,+deprecated-instance-without-locality,+deprecated-typeclasses-transparency-without-locality,unsupported-attributes -arg -w -arg -notation-overridden,-deprecated-hint-constr,-fragile-hint-constr,-native-compiler-disabled,-ambiguous-paths,-masking-absolute-name -arg -w -arg -deprecated-native-compiler-option -arg -native-compiler -arg ondemand bug_01.v
getting Makefilefg49ymn_.coq (/github/workspace/cwd/Makefilefg49ymn_.coq)
make -k -f Makefilefg49ymn_.coq KEEP_ERROR=1 bug_01.glob
getting bug_01.glob (/github/workspace/cwd/bug_01.glob)
getting bug_01.glob (/github/workspace/cwd/bug_01.glob)
[92m
Succeeded in normalizing Requires.[0m

Now, I will attempt to split up [Require] statements...
getting /github/workspace/cwd/bug_01.v
NOTE: The file /github/workspace/cwd/bug_01.v is very new (1668833528, 0 seconds old), delaying until it's a bit older
getting /github/workspace/cwd/bug_01.glob
getting /github/workspace/cwd/bug_01.glob

No Requires to split.

In order to efficiently manipulate the file, I have to break it into statements.  I will attempt to do this by matching on periods.
[92m
Splitting successful.[0m

I will now attempt to remove any lines after the line which generates the error.

No lines to trim.

In order to efficiently manipulate the file, I have to break it into definitions.  I will now attempt to do this.
Sending statements to coqtop...
Done.  Splitting to definitions...
[92m
Splitting to definitions successful.[0m

I will now attempt to remove goals ending in [Abort.]
[92m
Aborted removal successful.[0m

I will now attempt to remove unused Ltacs
[92m
Ltac removal successful.[0m

I will now attempt to remove unused definitions
[92m
Definition removal successful.[0m

I will now attempt to remove unused non-instance, non-canonical structure definitions
[92m
Non-instance definition removal successful.[0m

I will now attempt to remove unused variables
[92m
Variable removal successful.[0m

I will now attempt to remove unused contexts
[92m
Context removal successful.[0m

I will now attempt to replace Qed Obligation with Admit Obligations
[92m
Admitting Qed Obligations successful.[0m
Failed to do everything at once; trying one at a time.
Admitting Qed Obligations unsuccessful.
No successful changes.

I will now attempt to replace Qeds with Admitteds
[92m
Admitting Qeds successful.[0m
Failed to do everything at once; trying one at a time.
Admitting Qeds unsuccessful.
No successful changes.

I will now attempt to replace Qeds with admit. Defined.
[92m
Admitting Qeds successful.[0m
Failed to do everything at once; trying one at a time.
Admitting Qeds unsuccessful.
No successful changes.

I will now attempt to remove goals ending in [Abort.]
[92m
Aborted removal successful.[0m

I will now attempt to remove unused Ltacs
[92m
Ltac removal successful.[0m

I will now attempt to remove unused definitions
[92m
Definition removal successful.[0m

I will now attempt to remove unused non-instance, non-canonical structure definitions
[92m
Non-instance definition removal successful.[0m

I will now attempt to remove unused variables
[92m
Variable removal successful.[0m

I will now attempt to remove unused contexts
[92m
Context removal successful.[0m

I will now attempt to admit [abstract ...]s
[92m
Admitting [abstract ...] successful.[0m
[92m
Admitting [abstract ...] successful.[0m
Admitting [abstract ...] unsuccessful.
Admitting [abstract ...] unsuccessful.

I will now attempt to remove goals ending in [Abort.]
[92m
Aborted removal successful.[0m

I will now attempt to remove unused Ltacs
[92m
Ltac removal successful.[0m

I will now attempt to remove unused definitions
[92m
Definition removal successful.[0m

I will now attempt to remove unused non-instance, non-canonical structure definitions
[92m
Non-instance definition removal successful.[0m

I will now attempt to remove unused variables
[92m
Variable removal successful.[0m

I will now attempt to remove unused contexts
[92m
Context removal successful.[0m

I will now attempt to replace Obligation with Admit Obligations
[92m
Admitting Obligations successful.[0m
Failed to do everything at once; trying one at a time.
Admitting Obligations unsuccessful.
No successful changes.

I will now attempt to admit lemmas with Admitted
[92m
Admitting lemmas successful.[0m
Failed to do everything at once; trying one at a time.
Admitting lemmas unsuccessful.
No successful changes.

I will now attempt to admit definitions with Admitted
[92m
Admitting definitions successful.[0m
Failed to do everything at once; trying one at a time.
Admitting definitions unsuccessful.
No successful changes.

I will now attempt to admit lemmas with admit. Defined
[92m
Admitting lemmas successful.[0m
Failed to do everything at once; trying one at a time.
Admitting lemmas unsuccessful.
No successful changes.

I will now attempt to admit definitions with admit. Defined
[92m
Admitting definitions successful.[0m
Failed to do everything at once; trying one at a time.
Admitting definitions unsuccessful.
No successful changes.

I will now attempt to export modules
Module exportation unsuccessful.

I will now attempt to split imports and exports
Import/Export splitting unsuccessful.

I will now attempt to split := definitions
One-line definition splitting unsuccessful.

I will now attempt to remove all lines, one at a time
Line removal unsuccessful.

I will now attempt to remove goals ending in [Abort.]
[92m
Aborted removal successful.[0m

I will now attempt to remove unused Ltacs
[92m
Ltac removal successful.[0m

I will now attempt to remove unused definitions
[92m
Definition removal successful.[0m

I will now attempt to remove unused non-instance, non-canonical structure definitions
[92m
Non-instance definition removal successful.[0m

I will now attempt to remove unused variables
[92m
Variable removal successful.[0m

I will now attempt to remove unused contexts
[92m
Context removal successful.[0m

I will now attempt to remove empty sections

No empty sections to remove.

Now, I will attempt to strip repeated newlines and trailing spaces from this file...

No strippable newlines or spaces.
```
</details>

If you have any comments on your experience of the minimizer, please share them in a reply (possibly tagging `@JasonGross`).
If you believe there's a bug in the bug minimizer, please report it on [the bug minimizer issue tracker](https://github.com/JasonGross/coq-tools/issues).

_Originally posted by @coqbot in https://github.com/coq-community/run-coq-bug-minimizer/issues/2#issuecomment-1320801231_

--- a/dev/ci/user-overlays/16858-SkySkimmer-revert-broken-fiat.sh
+++ b/dev/ci/user-overlays/16858-SkySkimmer-revert-broken-fiat.sh
@@ -1,0 +1,1 @@
+overlay coq_dpdgraph https://github.com/SkySkimmer/coq-dpdgraph revert-broken-fiat 16858

--- a/kernel/safe_typing.ml
+++ b/kernel/safe_typing.ml
@@ -1243,13 +1243,13 @@ let add_include me is_module inl senv =
   let resolver, str = compute_sign sign resolver in
   let senv = update_resolver (Mod_subst.add_delta_resolver resolver) senv in
   let add senv ((l,elem) as field) =
-    let field, new_name = match elem with
+    let new_name = match elem with
       | SFBconst _ ->
-        field, C (Mod_subst.constant_of_delta_kn resolver (KerName.make mp_sup l))
+        C (Mod_subst.constant_of_delta_kn resolver (KerName.make mp_sup l))
       | SFBmind _ ->
-        field, I (Mod_subst.mind_of_delta_kn resolver (KerName.make mp_sup l))
-      | SFBmodule mb -> (l, SFBmodule {mb with mod_delta = Mod_subst.add_delta_resolver resolver mb.mod_delta}), M
-      | SFBmodtype _ -> (l, SFBmodtype {mb with mod_delta = Mod_subst.add_delta_resolver resolver mb.mod_delta}), MT
+        I (Mod_subst.mind_of_delta_kn resolver (KerName.make mp_sup l))
+      | SFBmodule _ -> M
+      | SFBmodtype _ -> MT
     in
     add_field field new_name senv
   in

--- a/test-suite/bugs/bug_16834.v
+++ b/test-suite/bugs/bug_16834.v
@@ -1,0 +1,9 @@
+Module Type DagDefinitions.
+End DagDefinitions.
+Module DagExtraDefinitions (dag : DagDefinitions).
+
+  Module Type EagerDagDefinitions.
+  End EagerDagDefinitions.
+End DagExtraDefinitions.
+Include DagExtraDefinitions.
+Declare Module eager : EagerDagDefinitions.

--- a/test-suite/output/Search.out
+++ b/test-suite/output/Search.out
@@ -20,105 +20,105 @@ le_sind:
   forall n0 : nat, n <= n0 -> P n0
 false: bool
 true: bool
-negb: bool -> bool
-eq_true: bool -> Prop
 is_true: bool -> Prop
-implb: bool -> bool -> bool
+eq_true: bool -> Prop
+negb: bool -> bool
+xorb: bool -> bool -> bool
 andb: bool -> bool -> bool
 orb: bool -> bool -> bool
-xorb: bool -> bool -> bool
-Nat.even: nat -> bool
+implb: bool -> bool -> bool
 Nat.odd: nat -> bool
+Nat.even: nat -> bool
 BoolSpec: Prop -> Prop -> bool -> Prop
-Nat.testbit: nat -> nat -> bool
+Number.number_beq: Number.number -> Number.number -> bool
 Nat.eqb: nat -> nat -> bool
-Hexadecimal.signed_int_beq:
-  Hexadecimal.signed_int -> Hexadecimal.signed_int -> bool
-Hexadecimal.uint_beq: Hexadecimal.uint -> Hexadecimal.uint -> bool
+Nat.testbit: nat -> nat -> bool
+Decimal.decimal_beq: Decimal.decimal -> Decimal.decimal -> bool
+Number.uint_beq: Number.uint -> Number.uint -> bool
 Decimal.uint_beq: Decimal.uint -> Decimal.uint -> bool
 Hexadecimal.hexadecimal_beq:
   Hexadecimal.hexadecimal -> Hexadecimal.hexadecimal -> bool
-Nat.leb: nat -> nat -> bool
-Decimal.decimal_beq: Decimal.decimal -> Decimal.decimal -> bool
-Decimal.signed_int_beq: Decimal.signed_int -> Decimal.signed_int -> bool
-Nat.ltb: nat -> nat -> bool
-Number.number_beq: Number.number -> Number.number -> bool
-Number.uint_beq: Number.uint -> Number.uint -> bool
 Number.signed_int_beq: Number.signed_int -> Number.signed_int -> bool
+Hexadecimal.uint_beq: Hexadecimal.uint -> Hexadecimal.uint -> bool
+Nat.ltb: nat -> nat -> bool
+Nat.leb: nat -> nat -> bool
+Hexadecimal.signed_int_beq:
+  Hexadecimal.signed_int -> Hexadecimal.signed_int -> bool
+Decimal.signed_int_beq: Decimal.signed_int -> Decimal.signed_int -> bool
 Nat.bitwise: (bool -> bool -> bool) -> nat -> nat -> nat -> nat
-eq_true_rec:
-  forall P : bool -> Set, P true -> forall b : bool, eq_true b -> P b
+eq_true_rec_r:
+  forall (P : bool -> Set) [b : bool], P b -> eq_true b -> P true
+eq_true_ind:
+  forall P : bool -> Prop, P true -> forall b : bool, eq_true b -> P b
 bool_rec: forall P : bool -> Set, P true -> P false -> forall b : bool, P b
 eq_true_rect_r:
   forall (P : bool -> Type) [b : bool], P b -> eq_true b -> P true
-bool_sind:
-  forall P : bool -> SProp, P true -> P false -> forall b : bool, P b
-eq_true_sind:
-  forall P : bool -> SProp, P true -> forall b : bool, eq_true b -> P b
-eq_true_ind:
-  forall P : bool -> Prop, P true -> forall b : bool, eq_true b -> P b
-bool_ind: forall P : bool -> Prop, P true -> P false -> forall b : bool, P b
-eq_true_rec_r:
-  forall (P : bool -> Set) [b : bool], P b -> eq_true b -> P true
-eq_true_rect:
-  forall P : bool -> Type, P true -> forall b : bool, eq_true b -> P b
+bool_rect: forall P : bool -> Type, P true -> P false -> forall b : bool, P b
 eq_true_ind_r:
   forall (P : bool -> Prop) [b : bool], P b -> eq_true b -> P true
-bool_rect: forall P : bool -> Type, P true -> P false -> forall b : bool, P b
-Decimal.internal_signed_int_dec_lb:
-  forall x y : Decimal.signed_int, x = y -> Decimal.signed_int_beq x y = true
-Decimal.internal_decimal_dec_lb:
-  forall x y : Decimal.decimal, x = y -> Decimal.decimal_beq x y = true
-Number.internal_uint_dec_lb1:
-  forall x y : Number.uint, x = y -> Number.uint_beq x y = true
-Number.internal_number_dec_bl:
-  forall x y : Number.number, Number.number_beq x y = true -> x = y
-Decimal.internal_decimal_dec_bl:
-  forall x y : Decimal.decimal, Decimal.decimal_beq x y = true -> x = y
-Number.internal_signed_int_dec_lb1:
-  forall x y : Number.signed_int, x = y -> Number.signed_int_beq x y = true
-Hexadecimal.internal_signed_int_dec_bl0:
-  forall x y : Hexadecimal.signed_int,
-  Hexadecimal.signed_int_beq x y = true -> x = y
-Decimal.internal_signed_int_dec_bl:
-  forall x y : Decimal.signed_int, Decimal.signed_int_beq x y = true -> x = y
+eq_true_rect:
+  forall P : bool -> Type, P true -> forall b : bool, eq_true b -> P b
+bool_sind:
+  forall P : bool -> SProp, P true -> P false -> forall b : bool, P b
+eq_true_rec:
+  forall P : bool -> Set, P true -> forall b : bool, eq_true b -> P b
+bool_ind: forall P : bool -> Prop, P true -> P false -> forall b : bool, P b
+eq_true_sind:
+  forall P : bool -> SProp, P true -> forall b : bool, eq_true b -> P b
+Number.internal_uint_dec_bl1:
+  forall x y : Number.uint, Number.uint_beq x y = true -> x = y
 Hexadecimal.internal_hexadecimal_dec_lb:
   forall x y : Hexadecimal.hexadecimal,
   x = y -> Hexadecimal.hexadecimal_beq x y = true
-Byte.to_bits:
-  Byte.byte ->
-  bool * (bool * (bool * (bool * (bool * (bool * (bool * bool))))))
-Number.internal_number_dec_lb:
-  forall x y : Number.number, x = y -> Number.number_beq x y = true
-Hexadecimal.internal_hexadecimal_dec_bl:
-  forall x y : Hexadecimal.hexadecimal,
-  Hexadecimal.hexadecimal_beq x y = true -> x = y
-Number.internal_signed_int_dec_bl1:
-  forall x y : Number.signed_int, Number.signed_int_beq x y = true -> x = y
 Hexadecimal.internal_signed_int_dec_lb0:
   forall x y : Hexadecimal.signed_int,
   x = y -> Hexadecimal.signed_int_beq x y = true
+Number.internal_number_dec_lb:
+  forall x y : Number.number, x = y -> Number.number_beq x y = true
+Decimal.internal_decimal_dec_lb:
+  forall x y : Decimal.decimal, x = y -> Decimal.decimal_beq x y = true
+Hexadecimal.internal_signed_int_dec_bl0:
+  forall x y : Hexadecimal.signed_int,
+  Hexadecimal.signed_int_beq x y = true -> x = y
+Number.internal_signed_int_dec_lb1:
+  forall x y : Number.signed_int, x = y -> Number.signed_int_beq x y = true
+Number.internal_signed_int_dec_bl1:
+  forall x y : Number.signed_int, Number.signed_int_beq x y = true -> x = y
+Hexadecimal.internal_hexadecimal_dec_bl:
+  forall x y : Hexadecimal.hexadecimal,
+  Hexadecimal.hexadecimal_beq x y = true -> x = y
+Number.internal_uint_dec_lb1:
+  forall x y : Number.uint, x = y -> Number.uint_beq x y = true
+Decimal.internal_signed_int_dec_bl:
+  forall x y : Decimal.signed_int, Decimal.signed_int_beq x y = true -> x = y
+Decimal.internal_signed_int_dec_lb:
+  forall x y : Decimal.signed_int, x = y -> Decimal.signed_int_beq x y = true
+Number.internal_number_dec_bl:
+  forall x y : Number.number, Number.number_beq x y = true -> x = y
 Byte.of_bits:
   bool * (bool * (bool * (bool * (bool * (bool * (bool * bool)))))) ->
   Byte.byte
-Number.internal_uint_dec_bl1:
-  forall x y : Number.uint, Number.uint_beq x y = true -> x = y
-Decimal.internal_uint_dec_bl:
-  forall x : Decimal.uint,
-  (fun x0 : Decimal.uint =>
-   forall y : Decimal.uint, Decimal.uint_beq x0 y = true -> x0 = y) x
+Byte.to_bits:
+  Byte.byte ->
+  bool * (bool * (bool * (bool * (bool * (bool * (bool * bool))))))
+Decimal.internal_decimal_dec_bl:
+  forall x y : Decimal.decimal, Decimal.decimal_beq x y = true -> x = y
 Hexadecimal.internal_uint_dec_bl0:
   forall x : Hexadecimal.uint,
   (fun x0 : Hexadecimal.uint =>
    forall y : Hexadecimal.uint, Hexadecimal.uint_beq x0 y = true -> x0 = y) x
-Hexadecimal.internal_uint_dec_lb0:
-  forall x : Hexadecimal.uint,
-  (fun x0 : Hexadecimal.uint =>
-   forall y : Hexadecimal.uint, x0 = y -> Hexadecimal.uint_beq x0 y = true) x
+Decimal.internal_uint_dec_bl:
+  forall x : Decimal.uint,
+  (fun x0 : Decimal.uint =>
+   forall y : Decimal.uint, Decimal.uint_beq x0 y = true -> x0 = y) x
 Decimal.internal_uint_dec_lb:
   forall x : Decimal.uint,
   (fun x0 : Decimal.uint =>
    forall y : Decimal.uint, x0 = y -> Decimal.uint_beq x0 y = true) x
+Hexadecimal.internal_uint_dec_lb0:
+  forall x : Hexadecimal.uint,
+  (fun x0 : Hexadecimal.uint =>
+   forall y : Hexadecimal.uint, x0 = y -> Hexadecimal.uint_beq x0 y = true) x
 andb_prop: forall a b : bool, (a && b)%bool = true -> a = true /\ b = true
 andb_true_intro:
   forall [b1 b2 : bool], b1 = true /\ b2 = true -> (b1 && b2)%bool = true
@@ -138,15 +138,15 @@ bool_choice:
   forall [S : Set] [R1 R2 : S -> Prop],
   (forall x : S, {R1 x} + {R2 x}) ->
   {f : S -> bool | forall x : S, f x = true /\ R1 x \/ f x = false /\ R2 x}
-plus_n_O: forall n : nat, n = n + 0
-plus_O_n: forall n : nat, 0 + n = n
 mult_n_O: forall n : nat, 0 = n * 0
-pred_Sn: forall n : nat, n = Nat.pred (S n)
+plus_O_n: forall n : nat, 0 + n = n
+plus_n_O: forall n : nat, n = n + 0
 n_Sn: forall n : nat, n <> S n
+pred_Sn: forall n : nat, n = Nat.pred (S n)
 O_S: forall n : nat, 0 <> S n
-eq_add_S: forall n m : nat, S n = S m -> n = m
-eq_S: forall x y : nat, x = y -> S x = S y
 f_equal_pred: forall x y : nat, x = y -> Nat.pred x = Nat.pred y
+eq_S: forall x y : nat, x = y -> S x = S y
+eq_add_S: forall n m : nat, S n = S m -> n = m
 min_r: forall n m : nat, m <= n -> Nat.min n m = m
 min_l: forall n m : nat, n <= m -> Nat.min n m = n
 max_r: forall n m : nat, n <= m -> Nat.max n m = m
@@ -156,57 +156,57 @@ plus_n_Sm: forall n m : nat, S (n + m) = n + S m
 f_equal_nat: forall (B : Type) (f : nat -> B) (x y : nat), x = y -> f x = f y
 not_eq_S: forall n m : nat, n <> m -> S n <> S m
 mult_n_Sm: forall n m : nat, n * m + n = n * S m
-f_equal2_mult:
-  forall x1 y1 x2 y2 : nat, x1 = y1 -> x2 = y2 -> x1 * x2 = y1 * y2
 f_equal2_plus:
   forall x1 y1 x2 y2 : nat, x1 = y1 -> x2 = y2 -> x1 + x2 = y1 + y2
+f_equal2_mult:
+  forall x1 y1 x2 y2 : nat, x1 = y1 -> x2 = y2 -> x1 * x2 = y1 * y2
 f_equal2_nat:
   forall (B : Type) (f : nat -> nat -> B) (x1 y1 x2 y2 : nat),
   x1 = y1 -> x2 = y2 -> f x1 x2 = f y1 y2
-Number.internal_uint_dec_lb1:
-  forall x y : Number.uint, x = y -> Number.uint_beq x y = true
-Number.internal_uint_dec_bl1:
-  forall x y : Number.uint, Number.uint_beq x y = true -> x = y
-Decimal.internal_signed_int_dec_lb:
-  forall x y : Decimal.signed_int, x = y -> Decimal.signed_int_beq x y = true
-Decimal.internal_signed_int_dec_bl:
-  forall x y : Decimal.signed_int, Decimal.signed_int_beq x y = true -> x = y
 Number.internal_number_dec_lb:
   forall x y : Number.number, x = y -> Number.number_beq x y = true
-Decimal.internal_decimal_dec_bl:
-  forall x y : Decimal.decimal, Decimal.decimal_beq x y = true -> x = y
 Number.internal_signed_int_dec_lb1:
   forall x y : Number.signed_int, x = y -> Number.signed_int_beq x y = true
 Number.internal_number_dec_bl:
   forall x y : Number.number, Number.number_beq x y = true -> x = y
-Decimal.internal_decimal_dec_lb:
-  forall x y : Decimal.decimal, x = y -> Decimal.decimal_beq x y = true
-Hexadecimal.internal_signed_int_dec_bl0:
-  forall x y : Hexadecimal.signed_int,
-  Hexadecimal.signed_int_beq x y = true -> x = y
 Hexadecimal.internal_hexadecimal_dec_lb:
   forall x y : Hexadecimal.hexadecimal,
   x = y -> Hexadecimal.hexadecimal_beq x y = true
 Number.internal_signed_int_dec_bl1:
   forall x y : Number.signed_int, Number.signed_int_beq x y = true -> x = y
+Number.internal_uint_dec_lb1:
+  forall x y : Number.uint, x = y -> Number.uint_beq x y = true
+Number.internal_uint_dec_bl1:
+  forall x y : Number.uint, Number.uint_beq x y = true -> x = y
+Decimal.internal_decimal_dec_lb:
+  forall x y : Decimal.decimal, x = y -> Decimal.decimal_beq x y = true
 Hexadecimal.internal_hexadecimal_dec_bl:
   forall x y : Hexadecimal.hexadecimal,
   Hexadecimal.hexadecimal_beq x y = true -> x = y
 Hexadecimal.internal_signed_int_dec_lb0:
   forall x y : Hexadecimal.signed_int,
   x = y -> Hexadecimal.signed_int_beq x y = true
-Hexadecimal.internal_uint_dec_lb0:
-  forall x : Hexadecimal.uint,
-  (fun x0 : Hexadecimal.uint =>
-   forall y : Hexadecimal.uint, x0 = y -> Hexadecimal.uint_beq x0 y = true) x
-Decimal.internal_uint_dec_lb:
-  forall x : Decimal.uint,
-  (fun x0 : Decimal.uint =>
-   forall y : Decimal.uint, x0 = y -> Decimal.uint_beq x0 y = true) x
+Hexadecimal.internal_signed_int_dec_bl0:
+  forall x y : Hexadecimal.signed_int,
+  Hexadecimal.signed_int_beq x y = true -> x = y
+Decimal.internal_signed_int_dec_lb:
+  forall x y : Decimal.signed_int, x = y -> Decimal.signed_int_beq x y = true
+Decimal.internal_decimal_dec_bl:
+  forall x y : Decimal.decimal, Decimal.decimal_beq x y = true -> x = y
+Decimal.internal_signed_int_dec_bl:
+  forall x y : Decimal.signed_int, Decimal.signed_int_beq x y = true -> x = y
 Decimal.internal_uint_dec_bl:
   forall x : Decimal.uint,
   (fun x0 : Decimal.uint =>
    forall y : Decimal.uint, Decimal.uint_beq x0 y = true -> x0 = y) x
+Decimal.internal_uint_dec_lb:
+  forall x : Decimal.uint,
+  (fun x0 : Decimal.uint =>
+   forall y : Decimal.uint, x0 = y -> Decimal.uint_beq x0 y = true) x
+Hexadecimal.internal_uint_dec_lb0:
+  forall x : Hexadecimal.uint,
+  (fun x0 : Hexadecimal.uint =>
+   forall y : Hexadecimal.uint, x0 = y -> Hexadecimal.uint_beq x0 y = true) x
 Hexadecimal.internal_uint_dec_bl0:
   forall x : Hexadecimal.uint,
   (fun x0 : Hexadecimal.uint =>
@@ -218,27 +218,18 @@ bool_choice:
   forall [S : Set] [R1 R2 : S -> Prop],
   (forall x : S, {R1 x} + {R2 x}) ->
   {f : S -> bool | forall x : S, f x = true /\ R1 x \/ f x = false /\ R2 x}
+Number.internal_number_dec_lb:
+  forall x y : Number.number, x = y -> Number.number_beq x y = true
+Number.internal_number_dec_bl:
+  forall x y : Number.number, Number.number_beq x y = true -> x = y
+Number.internal_signed_int_dec_lb1:
+  forall x y : Number.signed_int, x = y -> Number.signed_int_beq x y = true
+Number.internal_signed_int_dec_bl1:
+  forall x y : Number.signed_int, Number.signed_int_beq x y = true -> x = y
 Number.internal_uint_dec_lb1:
   forall x y : Number.uint, x = y -> Number.uint_beq x y = true
 Number.internal_uint_dec_bl1:
   forall x y : Number.uint, Number.uint_beq x y = true -> x = y
-Decimal.internal_signed_int_dec_lb:
-  forall x y : Decimal.signed_int, x = y -> Decimal.signed_int_beq x y = true
-Decimal.internal_signed_int_dec_bl:
-  forall x y : Decimal.signed_int, Decimal.signed_int_beq x y = true -> x = y
-Number.internal_number_dec_lb:
-  forall x y : Number.number, x = y -> Number.number_beq x y = true
-Decimal.internal_decimal_dec_bl:
-  forall x y : Decimal.decimal, Decimal.decimal_beq x y = true -> x = y
-Number.internal_signed_int_dec_lb1:
-  forall x y : Number.signed_int, x = y -> Number.signed_int_beq x y = true
-Number.internal_number_dec_bl:
-  forall x y : Number.number, Number.number_beq x y = true -> x = y
-Decimal.internal_decimal_dec_lb:
-  forall x y : Decimal.decimal, x = y -> Decimal.decimal_beq x y = true
-Hexadecimal.internal_signed_int_dec_bl0:
-  forall x y : Hexadecimal.signed_int,
-  Hexadecimal.signed_int_beq x y = true -> x = y
 Hexadecimal.internal_hexadecimal_dec_lb:
   forall x y : Hexadecimal.hexadecimal,
   x = y -> Hexadecimal.hexadecimal_beq x y = true
@@ -248,27 +239,36 @@ Hexadecimal.internal_hexadecimal_dec_bl:
 Hexadecimal.internal_signed_int_dec_lb0:
   forall x y : Hexadecimal.signed_int,
   x = y -> Hexadecimal.signed_int_beq x y = true
-Number.internal_signed_int_dec_bl1:
-  forall x y : Number.signed_int, Number.signed_int_beq x y = true -> x = y
-Decimal.internal_uint_dec_bl:
-  forall x : Decimal.uint,
-  (fun x0 : Decimal.uint =>
-   forall y : Decimal.uint, Decimal.uint_beq x0 y = true -> x0 = y) x
+Hexadecimal.internal_signed_int_dec_bl0:
+  forall x y : Hexadecimal.signed_int,
+  Hexadecimal.signed_int_beq x y = true -> x = y
+Decimal.internal_decimal_dec_lb:
+  forall x y : Decimal.decimal, x = y -> Decimal.decimal_beq x y = true
+Decimal.internal_decimal_dec_bl:
+  forall x y : Decimal.decimal, Decimal.decimal_beq x y = true -> x = y
+Decimal.internal_signed_int_dec_lb:
+  forall x y : Decimal.signed_int, x = y -> Decimal.signed_int_beq x y = true
+Decimal.internal_signed_int_dec_bl:
+  forall x y : Decimal.signed_int, Decimal.signed_int_beq x y = true -> x = y
 Hexadecimal.internal_uint_dec_bl0:
   forall x : Hexadecimal.uint,
   (fun x0 : Hexadecimal.uint =>
    forall y : Hexadecimal.uint, Hexadecimal.uint_beq x0 y = true -> x0 = y) x
-Hexadecimal.internal_uint_dec_lb0:
-  forall x : Hexadecimal.uint,
-  (fun x0 : Hexadecimal.uint =>
-   forall y : Hexadecimal.uint, x0 = y -> Hexadecimal.uint_beq x0 y = true) x
+Decimal.internal_uint_dec_bl:
+  forall x : Decimal.uint,
+  (fun x0 : Decimal.uint =>
+   forall y : Decimal.uint, Decimal.uint_beq x0 y = true -> x0 = y) x
 Decimal.internal_uint_dec_lb:
   forall x : Decimal.uint,
   (fun x0 : Decimal.uint =>
    forall y : Decimal.uint, x0 = y -> Decimal.uint_beq x0 y = true) x
-andb_prop: forall a b : bool, (a && b)%bool = true -> a = true /\ b = true
+Hexadecimal.internal_uint_dec_lb0:
+  forall x : Hexadecimal.uint,
+  (fun x0 : Hexadecimal.uint =>
+   forall y : Hexadecimal.uint, x0 = y -> Hexadecimal.uint_beq x0 y = true) x
 andb_true_intro:
   forall [b1 b2 : bool], b1 = true /\ b2 = true -> (b1 && b2)%bool = true
+andb_prop: forall a b : bool, (a && b)%bool = true -> a = true /\ b = true
 andb_prop: forall a b : bool, (a && b)%bool = true -> a = true /\ b = true
 h: n <> newdef n
 h': newdef n <> n
@@ -302,8 +302,8 @@ and_assoc: forall A B C : Prop, (A /\ B) /\ C <-> A /\ B /\ C
 eq_trans_assoc:
   forall [A : Type] [x y z t : A] (e : x = y) (e' : y = z) (e'' : z = t),
   eq_trans e (eq_trans e' e'') = eq_trans (eq_trans e e') e''
-plus_n_O: forall n : nat, n = n + 0
 plus_O_n: forall n : nat, 0 + n = n
+plus_n_O: forall n : nat, n = n + 0
 plus_n_Sm: forall n m : nat, S (n + m) = n + S m
 plus_Sn_m: forall n m : nat, S n + m = S (n + m)
 mult_n_Sm: forall n m : nat, n * m + n = n * S m
@@ -316,67 +316,67 @@ nat_rect_plus:
     (nat_rect (fun _ : nat => A) x (fun _ : nat => f) m) 
     (fun _ : nat => f) n
 Nat.bitwise: (bool -> bool -> bool) -> nat -> nat -> nat -> nat
-Number.internal_uint_dec_bl1:
-  forall x y : Number.uint, Number.uint_beq x y = true -> x = y
-Byte.of_bits:
-  bool * (bool * (bool * (bool * (bool * (bool * (bool * bool)))))) ->
-  Byte.byte
-Decimal.internal_signed_int_dec_bl:
-  forall x y : Decimal.signed_int, Decimal.signed_int_beq x y = true -> x = y
 Number.internal_number_dec_bl:
   forall x y : Number.number, Number.number_beq x y = true -> x = y
-Decimal.internal_decimal_dec_bl:
-  forall x y : Decimal.decimal, Decimal.decimal_beq x y = true -> x = y
 Number.internal_signed_int_dec_bl1:
   forall x y : Number.signed_int, Number.signed_int_beq x y = true -> x = y
-Hexadecimal.internal_signed_int_dec_bl0:
-  forall x y : Hexadecimal.signed_int,
-  Hexadecimal.signed_int_beq x y = true -> x = y
+Number.internal_uint_dec_bl1:
+  forall x y : Number.uint, Number.uint_beq x y = true -> x = y
 Hexadecimal.internal_hexadecimal_dec_bl:
   forall x y : Hexadecimal.hexadecimal,
   Hexadecimal.hexadecimal_beq x y = true -> x = y
+Hexadecimal.internal_signed_int_dec_bl0:
+  forall x y : Hexadecimal.signed_int,
+  Hexadecimal.signed_int_beq x y = true -> x = y
+Decimal.internal_decimal_dec_bl:
+  forall x y : Decimal.decimal, Decimal.decimal_beq x y = true -> x = y
+Decimal.internal_signed_int_dec_bl:
+  forall x y : Decimal.signed_int, Decimal.signed_int_beq x y = true -> x = y
+Byte.of_bits:
+  bool * (bool * (bool * (bool * (bool * (bool * (bool * bool)))))) ->
+  Byte.byte
 Byte.to_bits_of_bits:
   forall
     b : bool * (bool * (bool * (bool * (bool * (bool * (bool * bool)))))),
   Byte.to_bits (Byte.of_bits b) = b
-Number.internal_uint_dec_lb1:
-  forall x y : Number.uint, x = y -> Number.uint_beq x y = true
-Byte.to_bits:
-  Byte.byte ->
-  bool * (bool * (bool * (bool * (bool * (bool * (bool * bool))))))
-Decimal.internal_decimal_dec_lb:
-  forall x y : Decimal.decimal, x = y -> Decimal.decimal_beq x y = true
-Decimal.internal_signed_int_dec_lb:
-  forall x y : Decimal.signed_int, x = y -> Decimal.signed_int_beq x y = true
-Hexadecimal.internal_signed_int_dec_lb0:
-  forall x y : Hexadecimal.signed_int,
-  x = y -> Hexadecimal.signed_int_beq x y = true
 Number.internal_number_dec_lb:
   forall x y : Number.number, x = y -> Number.number_beq x y = true
+Number.internal_uint_dec_lb1:
+  forall x y : Number.uint, x = y -> Number.uint_beq x y = true
+Number.internal_signed_int_dec_lb1:
+  forall x y : Number.signed_int, x = y -> Number.signed_int_beq x y = true
+Decimal.internal_signed_int_dec_lb:
+  forall x y : Decimal.signed_int, x = y -> Decimal.signed_int_beq x y = true
 Hexadecimal.internal_hexadecimal_dec_lb:
   forall x y : Hexadecimal.hexadecimal,
   x = y -> Hexadecimal.hexadecimal_beq x y = true
-Number.internal_signed_int_dec_lb1:
-  forall x y : Number.signed_int, x = y -> Number.signed_int_beq x y = true
+Hexadecimal.internal_signed_int_dec_lb0:
+  forall x y : Hexadecimal.signed_int,
+  x = y -> Hexadecimal.signed_int_beq x y = true
+Decimal.internal_decimal_dec_lb:
+  forall x y : Decimal.decimal, x = y -> Decimal.decimal_beq x y = true
+Byte.to_bits:
+  Byte.byte ->
+  bool * (bool * (bool * (bool * (bool * (bool * (bool * bool))))))
 Hexadecimal.internal_uint_dec_bl0:
   forall x : Hexadecimal.uint,
   (fun x0 : Hexadecimal.uint =>
    forall y : Hexadecimal.uint, Hexadecimal.uint_beq x0 y = true -> x0 = y) x
-Hexadecimal.internal_uint_dec_lb0:
-  forall x : Hexadecimal.uint,
-  (fun x0 : Hexadecimal.uint =>
-   forall y : Hexadecimal.uint, x0 = y -> Hexadecimal.uint_beq x0 y = true) x
-Decimal.internal_uint_dec_bl:
-  forall x : Decimal.uint,
-  (fun x0 : Decimal.uint =>
-   forall y : Decimal.uint, Decimal.uint_beq x0 y = true -> x0 = y) x
 Decimal.internal_uint_dec_lb:
   forall x : Decimal.uint,
   (fun x0 : Decimal.uint =>
    forall y : Decimal.uint, x0 = y -> Decimal.uint_beq x0 y = true) x
+Decimal.internal_uint_dec_bl:
+  forall x : Decimal.uint,
+  (fun x0 : Decimal.uint =>
+   forall y : Decimal.uint, Decimal.uint_beq x0 y = true -> x0 = y) x
+Hexadecimal.internal_uint_dec_lb0:
+  forall x : Hexadecimal.uint,
+  (fun x0 : Hexadecimal.uint =>
+   forall y : Hexadecimal.uint, x0 = y -> Hexadecimal.uint_beq x0 y = true) x
+andb_prop: forall a b : bool, (a && b)%bool = true -> a = true /\ b = true
 andb_true_intro:
   forall [b1 b2 : bool], b1 = true /\ b2 = true -> (b1 && b2)%bool = true
-andb_prop: forall a b : bool, (a && b)%bool = true -> a = true /\ b = true
 Byte.to_bits_of_bits:
   forall
     b : bool * (bool * (bool * (bool * (bool * (bool * (bool * bool)))))),
@@ -385,43 +385,43 @@ bool_choice:
   forall [S : Set] [R1 R2 : S -> Prop],
   (forall x : S, {R1 x} + {R2 x}) ->
   {f : S -> bool | forall x : S, f x = true /\ R1 x \/ f x = false /\ R2 x}
-Nat.zero: nat
 Nat.two: nat
+Nat.zero: nat
 Nat.one: nat
+Nat.succ: nat -> nat
+Nat.log2: nat -> nat
+Nat.sqrt: nat -> nat
+Nat.square: nat -> nat
 Nat.double: nat -> nat
 Nat.pred: nat -> nat
-Nat.square: nat -> nat
-Nat.succ: nat -> nat
-Nat.sqrt: nat -> nat
-Nat.log2: nat -> nat
-Nat.land: nat -> nat -> nat
-Nat.modulo: nat -> nat -> nat
-Nat.tail_mul: nat -> nat -> nat
-Nat.div: nat -> nat -> nat
-Nat.lor: nat -> nat -> nat
 Nat.ldiff: nat -> nat -> nat
+Nat.tail_mul: nat -> nat -> nat
+Nat.land: nat -> nat -> nat
+Nat.div: nat -> nat -> nat
+Nat.modulo: nat -> nat -> nat
+Nat.lor: nat -> nat -> nat
 Nat.lxor: nat -> nat -> nat
 Nat.of_hex_uint: Hexadecimal.uint -> nat
-Nat.of_num_uint: Number.uint -> nat
 Nat.of_uint: Decimal.uint -> nat
+Nat.of_num_uint: Number.uint -> nat
 length: forall [A : Type], list A -> nat
-plus_O_n: forall n : nat, 0 + n = n
 plus_n_O: forall n : nat, n = n + 0
+plus_O_n: forall n : nat, 0 + n = n
 plus_n_Sm: forall n m : nat, S (n + m) = n + S m
 plus_Sn_m: forall n m : nat, S n + m = S (n + m)
 mult_n_Sm: forall n m : nat, n * m + n = n * S m
-Nat.mul_comm: forall n m : nat, n * m = m * n
-Bool.xorb_comm: forall b b' : bool, xorb b b' = xorb b' b
 Nat.land_comm: forall a b : nat, Nat.land a b = Nat.land b a
 Nat.lor_comm: forall a b : nat, Nat.lor a b = Nat.lor b a
+Nat.lxor_comm: forall a b : nat, Nat.lxor a b = Nat.lxor b a
+Nat.lcm_comm: forall a b : nat, Nat.lcm a b = Nat.lcm b a
+Nat.min_comm: forall n m : nat, Nat.min n m = Nat.min m n
+Nat.gcd_comm: forall n m : nat, Nat.gcd n m = Nat.gcd m n
+Bool.xorb_comm: forall b b' : bool, xorb b b' = xorb b' b
+Nat.max_comm: forall n m : nat, Nat.max n m = Nat.max m n
+Nat.mul_comm: forall n m : nat, n * m = m * n
+Nat.add_comm: forall n m : nat, n + m = m + n
 Bool.orb_comm: forall b1 b2 : bool, (b1 || b2)%bool = (b2 || b1)%bool
 Bool.andb_comm: forall b1 b2 : bool, (b1 && b2)%bool = (b2 && b1)%bool
-Nat.min_comm: forall n m : nat, Nat.min n m = Nat.min m n
-Nat.add_comm: forall n m : nat, n + m = m + n
-Nat.lcm_comm: forall a b : nat, Nat.lcm a b = Nat.lcm b a
-Nat.lxor_comm: forall a b : nat, Nat.lxor a b = Nat.lxor b a
-Nat.gcd_comm: forall n m : nat, Nat.gcd n m = Nat.gcd m n
-Nat.max_comm: forall n m : nat, Nat.max n m = Nat.max m n
 Nat.eqb_sym: forall x y : nat, (x =? y) = (y =? x)
 Nat.bit0_eqb: forall a : nat, Nat.testbit a 0 = (a mod 2 =? 1)
 Nat.Div0.div_exact: forall a b : nat, a = b * (a / b) <-> a mod b = 0
@@ -447,23 +447,23 @@ Nat.testbit_false:
   forall a n : nat, Nat.testbit a n = false <-> (a / 2 ^ n) mod 2 = 0
 Nat.testbit_true:
   forall a n : nat, Nat.testbit a n = true <-> (a / 2 ^ n) mod 2 = 1
+iff_Symmetric: Symmetric iff
 iff_Reflexive: Reflexive iff
 impl_Reflexive: Reflexive Basics.impl
-iff_Symmetric: Symmetric iff
 eq_Symmetric: forall {A : Type}, Symmetric eq
 eq_Reflexive: forall {A : Type}, Reflexive eq
-PreOrder_Reflexive:
-  forall {A : Type} {R : Relation_Definitions.relation A},
-  PreOrder R -> Reflexive R
-PER_Symmetric:
-  forall {A : Type} {R : Relation_Definitions.relation A},
-  PER R -> Symmetric R
 Equivalence_Symmetric:
   forall {A : Type} {R : Relation_Definitions.relation A},
   Equivalence R -> Symmetric R
 Equivalence_Reflexive:
   forall {A : Type} {R : Relation_Definitions.relation A},
   Equivalence R -> Reflexive R
+PER_Symmetric:
+  forall {A : Type} {R : Relation_Definitions.relation A},
+  PER R -> Symmetric R
+PreOrder_Reflexive:
+  forall {A : Type} {R : Relation_Definitions.relation A},
+  PreOrder R -> Reflexive R
 reflexive_eq_dom_reflexive:
   forall {A B : Type} {R' : Relation_Definitions.relation B},
   Reflexive R' -> Reflexive (eq ==> R')%signature

--- a/test-suite/output/SearchPattern.out
+++ b/test-suite/output/SearchPattern.out
@@ -1,92 +1,92 @@
 false: bool
 true: bool
 negb: bool -> bool
-implb: bool -> bool -> bool
 xorb: bool -> bool -> bool
 andb: bool -> bool -> bool
 orb: bool -> bool -> bool
+implb: bool -> bool -> bool
 Nat.odd: nat -> bool
 Nat.even: nat -> bool
-Decimal.signed_int_beq: Decimal.signed_int -> Decimal.signed_int -> bool
-Nat.eqb: nat -> nat -> bool
+Number.uint_beq: Number.uint -> Number.uint -> bool
 Nat.testbit: nat -> nat -> bool
-Number.signed_int_beq: Number.signed_int -> Number.signed_int -> bool
+Nat.eqb: nat -> nat -> bool
 Hexadecimal.hexadecimal_beq:
   Hexadecimal.hexadecimal -> Hexadecimal.hexadecimal -> bool
-Number.number_beq: Number.number -> Number.number -> bool
-Nat.leb: nat -> nat -> bool
 Nat.ltb: nat -> nat -> bool
-Decimal.uint_beq: Decimal.uint -> Decimal.uint -> bool
+Nat.leb: nat -> nat -> bool
+Number.number_beq: Number.number -> Number.number -> bool
+Number.signed_int_beq: Number.signed_int -> Number.signed_int -> bool
 Hexadecimal.signed_int_beq:
   Hexadecimal.signed_int -> Hexadecimal.signed_int -> bool
-Number.uint_beq: Number.uint -> Number.uint -> bool
-Decimal.decimal_beq: Decimal.decimal -> Decimal.decimal -> bool
 Hexadecimal.uint_beq: Hexadecimal.uint -> Hexadecimal.uint -> bool
-O: nat
+Decimal.decimal_beq: Decimal.decimal -> Decimal.decimal -> bool
+Decimal.signed_int_beq: Decimal.signed_int -> Decimal.signed_int -> bool
+Decimal.uint_beq: Decimal.uint -> Decimal.uint -> bool
 Nat.two: nat
 Nat.zero: nat
 Nat.one: nat
-Nat.succ: nat -> nat
+O: nat
 Nat.double: nat -> nat
-Nat.log2: nat -> nat
-S: nat -> nat
-Nat.pred: nat -> nat
 Nat.sqrt: nat -> nat
 Nat.div2: nat -> nat
+Nat.log2: nat -> nat
+Nat.pred: nat -> nat
 Nat.square: nat -> nat
-Nat.tail_mul: nat -> nat -> nat
-Nat.div: nat -> nat -> nat
+S: nat -> nat
+Nat.succ: nat -> nat
 Nat.ldiff: nat -> nat -> nat
-Nat.pow: nat -> nat -> nat
+Nat.add: nat -> nat -> nat
+Nat.land: nat -> nat -> nat
+Nat.lxor: nat -> nat -> nat
+Nat.sub: nat -> nat -> nat
 Nat.mul: nat -> nat -> nat
-Nat.tail_add: nat -> nat -> nat
+Nat.tail_mul: nat -> nat -> nat
 Nat.max: nat -> nat -> nat
+Nat.tail_add: nat -> nat -> nat
+Nat.pow: nat -> nat -> nat
+Nat.min: nat -> nat -> nat
+Nat.modulo: nat -> nat -> nat
+Nat.div: nat -> nat -> nat
 Nat.lor: nat -> nat -> nat
 Nat.gcd: nat -> nat -> nat
-Nat.land: nat -> nat -> nat
-Nat.sub: nat -> nat -> nat
-Nat.lxor: nat -> nat -> nat
-Nat.modulo: nat -> nat -> nat
-Nat.min: nat -> nat -> nat
-Nat.add: nat -> nat -> nat
-Nat.of_uint: Decimal.uint -> nat
-Nat.of_num_uint: Number.uint -> nat
 Hexadecimal.nb_digits: Hexadecimal.uint -> nat
-Decimal.nb_digits: Decimal.uint -> nat
 Nat.of_hex_uint: Hexadecimal.uint -> nat
+Nat.of_num_uint: Number.uint -> nat
+Nat.of_uint: Decimal.uint -> nat
+Decimal.nb_digits: Decimal.uint -> nat
 Nat.tail_addmul: nat -> nat -> nat -> nat
-Nat.of_uint_acc: Decimal.uint -> nat -> nat
 Nat.of_hex_uint_acc: Hexadecimal.uint -> nat -> nat
+Nat.of_uint_acc: Decimal.uint -> nat -> nat
 Nat.sqrt_iter: nat -> nat -> nat -> nat -> nat
 Nat.log2_iter: nat -> nat -> nat -> nat -> nat
 length: forall [A : Type], list A -> nat
 Nat.bitwise: (bool -> bool -> bool) -> nat -> nat -> nat -> nat
+Nat.div2: nat -> nat
+Nat.sqrt: nat -> nat
+Nat.log2: nat -> nat
+Nat.double: nat -> nat
 S: nat -> nat
 Nat.square: nat -> nat
-Nat.sqrt: nat -> nat
-Nat.double: nat -> nat
-Nat.log2: nat -> nat
 Nat.succ: nat -> nat
-Nat.div2: nat -> nat
 Nat.pred: nat -> nat
-Nat.max: nat -> nat -> nat
-Nat.lxor: nat -> nat -> nat
-Nat.tail_add: nat -> nat -> nat
 Nat.land: nat -> nat -> nat
-Nat.sub: nat -> nat -> nat
-Nat.tail_mul: nat -> nat -> nat
+Nat.max: nat -> nat -> nat
+Nat.gcd: nat -> nat -> nat
 Nat.modulo: nat -> nat -> nat
 Nat.ldiff: nat -> nat -> nat
-Nat.add: nat -> nat -> nat
-Nat.lor: nat -> nat -> nat
-Nat.gcd: nat -> nat -> nat
-Nat.div: nat -> nat -> nat
-Nat.mul: nat -> nat -> nat
+Nat.tail_add: nat -> nat -> nat
 Nat.pow: nat -> nat -> nat
+Nat.lxor: nat -> nat -> nat
+Nat.div: nat -> nat -> nat
+Nat.lor: nat -> nat -> nat
+Nat.mul: nat -> nat -> nat
 Nat.min: nat -> nat -> nat
+Nat.add: nat -> nat -> nat
+Nat.sub: nat -> nat -> nat
+Nat.tail_mul: nat -> nat -> nat
 Nat.tail_addmul: nat -> nat -> nat -> nat
-Nat.of_hex_uint_acc: Hexadecimal.uint -> nat -> nat
 Nat.of_uint_acc: Decimal.uint -> nat -> nat
+Nat.of_hex_uint_acc: Hexadecimal.uint -> nat -> nat
 Nat.sqrt_iter: nat -> nat -> nat -> nat -> nat
 Nat.log2_iter: nat -> nat -> nat -> nat -> nat
 Nat.bitwise: (bool -> bool -> bool) -> nat -> nat -> nat -> nat

--- a/test-suite/output/Search_headconcl.out
+++ b/test-suite/output/Search_headconcl.out
@@ -1,50 +1,50 @@
 le_n: forall n : nat, n <= n
 le_0_n: forall n : nat, 0 <= n
 le_S: forall n m : nat, n <= m -> n <= S m
-le_S_n: forall n m : nat, S n <= S m -> n <= m
-le_n_S: forall n m : nat, n <= m -> S n <= S m
 le_pred: forall n m : nat, n <= m -> Nat.pred n <= Nat.pred m
+le_n_S: forall n m : nat, n <= m -> S n <= S m
+le_S_n: forall n m : nat, S n <= S m -> n <= m
 false: bool
 true: bool
 negb: bool -> bool
-implb: bool -> bool -> bool
 xorb: bool -> bool -> bool
 andb: bool -> bool -> bool
 orb: bool -> bool -> bool
+implb: bool -> bool -> bool
 Nat.odd: nat -> bool
 Nat.even: nat -> bool
-Decimal.signed_int_beq: Decimal.signed_int -> Decimal.signed_int -> bool
-Nat.eqb: nat -> nat -> bool
+Number.uint_beq: Number.uint -> Number.uint -> bool
 Nat.testbit: nat -> nat -> bool
-Number.signed_int_beq: Number.signed_int -> Number.signed_int -> bool
+Nat.eqb: nat -> nat -> bool
 Hexadecimal.hexadecimal_beq:
   Hexadecimal.hexadecimal -> Hexadecimal.hexadecimal -> bool
-Number.number_beq: Number.number -> Number.number -> bool
-Nat.leb: nat -> nat -> bool
 Nat.ltb: nat -> nat -> bool
-Decimal.uint_beq: Decimal.uint -> Decimal.uint -> bool
+Nat.leb: nat -> nat -> bool
+Number.number_beq: Number.number -> Number.number -> bool
+Number.signed_int_beq: Number.signed_int -> Number.signed_int -> bool
 Hexadecimal.signed_int_beq:
   Hexadecimal.signed_int -> Hexadecimal.signed_int -> bool
-Number.uint_beq: Number.uint -> Number.uint -> bool
-Decimal.decimal_beq: Decimal.decimal -> Decimal.decimal -> bool
 Hexadecimal.uint_beq: Hexadecimal.uint -> Hexadecimal.uint -> bool
-plus_n_O: forall n : nat, n = n + 0
-plus_O_n: forall n : nat, 0 + n = n
+Decimal.decimal_beq: Decimal.decimal -> Decimal.decimal -> bool
+Decimal.signed_int_beq: Decimal.signed_int -> Decimal.signed_int -> bool
+Decimal.uint_beq: Decimal.uint -> Decimal.uint -> bool
 mult_n_O: forall n : nat, 0 = n * 0
+plus_O_n: forall n : nat, 0 + n = n
+plus_n_O: forall n : nat, n = n + 0
 pred_Sn: forall n : nat, n = Nat.pred (S n)
+f_equal_pred: forall x y : nat, x = y -> Nat.pred x = Nat.pred y
 eq_add_S: forall n m : nat, S n = S m -> n = m
 eq_S: forall x y : nat, x = y -> S x = S y
-f_equal_pred: forall x y : nat, x = y -> Nat.pred x = Nat.pred y
+max_r: forall n m : nat, n <= m -> Nat.max n m = m
 max_l: forall n m : nat, m <= n -> Nat.max n m = n
 min_r: forall n m : nat, m <= n -> Nat.min n m = m
-max_r: forall n m : nat, n <= m -> Nat.max n m = m
 min_l: forall n m : nat, n <= m -> Nat.min n m = n
-plus_Sn_m: forall n m : nat, S n + m = S (n + m)
 plus_n_Sm: forall n m : nat, S (n + m) = n + S m
+plus_Sn_m: forall n m : nat, S n + m = S (n + m)
 mult_n_Sm: forall n m : nat, n * m + n = n * S m
-f_equal2_mult:
-  forall x1 y1 x2 y2 : nat, x1 = y1 -> x2 = y2 -> x1 * x2 = y1 * y2
 f_equal2_plus:
   forall x1 y1 x2 y2 : nat, x1 = y1 -> x2 = y2 -> x1 + x2 = y1 + y2
+f_equal2_mult:
+  forall x1 y1 x2 y2 : nat, x1 = y1 -> x2 = y2 -> x1 * x2 = y1 * y2
 h: newdef n
 h: P n

--- a/test-suite/output/bug_13244.out
+++ b/test-suite/output/bug_13244.out
@@ -1,9 +1,9 @@
 negbT: forall [b : bool], b = false -> ~~ b
-contraPN: forall [P : Prop] [b : bool], (b -> ~ P) -> P -> ~~ b
 contra_notN: forall [P : Prop] [b : bool], (b -> P) -> ~ P -> ~~ b
+contraPN: forall [P : Prop] [b : bool], (b -> ~ P) -> P -> ~~ b
+contraNN: forall [c b : bool], (c -> b) -> ~~ b -> ~~ c
 contraL: forall [c b : bool], (c -> ~~ b) -> b -> ~~ c
 contraTN: forall [c b : bool], (c -> ~~ b) -> b -> ~~ c
 contra: forall [c b : bool], (c -> b) -> ~~ b -> ~~ c
-contraNN: forall [c b : bool], (c -> b) -> ~~ b -> ~~ c
 introN: forall [P : Prop] [b : bool], reflect P b -> ~ P -> ~~ b
 contraFN: forall [c b : bool], (c -> b) -> b = false -> ~~ c

--- a/vernac/declaremods.ml
+++ b/vernac/declaremods.ml
@@ -1130,6 +1130,25 @@ let import_modules ~export mpl =
 let import_module f ~export mp =
   import_modules ~export [f,mp]
 
+(** {6 Iterators} *)
+
+let iter_all_segments f =
+  let rec apply_obj prefix obj = match obj with
+    | IncludeObject aobjs ->
+      let objs = expand_aobjs aobjs in
+      List.iter (apply_obj prefix) objs
+    | _ -> f prefix obj
+  in
+  let apply_mod_obj _ modobjs =
+    let prefix = modobjs.module_prefix in
+    List.iter (apply_obj prefix) modobjs.module_substituted_objects;
+    List.iter (apply_obj prefix) modobjs.module_keep_objects
+  in
+  let apply_nodes (node, os) = List.iter (fun o -> f (Lib.node_prefix node) o) os in
+  MPmap.iter apply_mod_obj (ModObjs.all ());
+  List.iter apply_nodes (Lib.contents ())
+
+
 (** {6 Some types used to shorten declaremods.mli} *)
 
 type module_params = (lident list * (Constrexpr.module_ast * inline)) list

--- a/vernac/declaremods.mli
+++ b/vernac/declaremods.mli
@@ -108,6 +108,14 @@ val import_modules : export:Lib.export_flag -> (Libobject.open_filter * ModPath.
 val declare_include : (Constrexpr.module_ast * inline) list -> unit
 
 (** {6 ... } *)
+(** [iter_all_segments] iterate over all segments, the modules'
+    segments first and then the current segment. Modules are presented
+    in an arbitrary order. The given function is applied to all leaves
+    (together with their section path). *)
+
+val iter_all_segments :
+  (Nametab.object_prefix -> Libobject.t -> unit) -> unit
+
 
 val debug_print_modtab : unit -> Pp.t
 

--- a/vernac/search.ml
+++ b/vernac/search.ml
@@ -13,6 +13,7 @@ open Util
 open Names
 open Constr
 open Declarations
+open Libobject
 open Environ
 open Pattern
 open Libnames
@@ -63,31 +64,51 @@ let iter_constructors indsp u fn env nconstr =
     fn (GlobRef.ConstructRef (indsp, i)) None env typ
   done
 
+(* FIXME: this is a Libobject hack that should be replaced with a proper
+   registration mechanism. *)
+module DynHandle = Libobject.Dyn.Map(struct type 'a t = 'a -> unit end)
+
+let handle h (Libobject.Dyn.Dyn (tag, o)) = match DynHandle.find tag h with
+| f -> f o
+| exception Not_found -> ()
+
 (* General search over declarations *)
 let generic_search env (fn : GlobRef.t -> Decls.logical_kind option -> env -> constr -> unit) =
-  let iter_var d = fn (GlobRef.VarRef (NamedDecl.get_id d)) None env (NamedDecl.get_type d) in
-  let iter_constant cst _ =
-    let gr = GlobRef.ConstRef cst in
-    let (typ, _) = Typeops.type_of_global_in_context (Global.env ()) gr in
-    let kind = Dumpglob.constant_kind cst in
-    fn gr (Some kind) env typ
+  List.iter (fun d -> fn (GlobRef.VarRef (NamedDecl.get_id d)) None env (NamedDecl.get_type d))
+    (Environ.named_context env);
+  let iter_obj prefix lobj = match lobj with
+    | AtomicObject o ->
+      let handler =
+        DynHandle.add Declare.Internal.Constant.tag begin fun (id,obj) ->
+          let kn = KerName.make prefix.Nametab.obj_mp (Label.of_id id) in
+          let cst = Global.constant_of_delta_kn kn in
+          let gr = GlobRef.ConstRef cst in
+          let (typ, _) = Typeops.type_of_global_in_context (Global.env ()) gr in
+          let kind = Declare.Internal.Constant.kind obj in
+          fn gr (Some kind) env typ
+        end @@
+        DynHandle.add DeclareInd.Internal.objInductive begin fun (id,_) ->
+          let kn = KerName.make prefix.Nametab.obj_mp (Label.of_id id) in
+          let mind = Global.mind_of_delta_kn kn in
+          let mib = Global.lookup_mind mind in
+          let iter_packet i mip =
+            let ind = (mind, i) in
+            let u = Univ.make_abstract_instance (Declareops.inductive_polymorphic_context mib) in
+            let i = (ind, u) in
+            let typ = Inductiveops.type_of_inductive env i in
+            let () = fn (GlobRef.IndRef ind) None env typ in
+            let len = Array.length mip.mind_user_lc in
+            iter_constructors ind u fn env len
+          in
+          Array.iteri iter_packet mib.mind_packets
+        end @@
+        DynHandle.empty
+      in
+      handle handler o
+    | _ -> ()
   in
-  let iter_inductive mind (mib,_) =
-    let iter_packet i mip =
-      let ind = (mind, i) in
-      let u = Univ.make_abstract_instance (Declareops.inductive_polymorphic_context mib) in
-      let i = (ind, u) in
-      let typ = Inductiveops.type_of_inductive env i in
-      let () = fn (GlobRef.IndRef ind) None env typ in
-      let len = Array.length mip.mind_user_lc in
-      iter_constructors ind u fn env len
-    in
-    Array.iteri iter_packet mib.mind_packets
-  in
-  List.iter iter_var (Environ.named_context env);
-  let globals = Environ.Globals.view env.env_globals in
-  Cmap_env.iter iter_constant globals.constants;
-  Mindmap_env.iter iter_inductive globals.inductives
+  try Declaremods.iter_all_segments iter_obj
+  with Not_found -> ()
 
 (** This module defines a preference on constrs in the form of a
     [compare] function (preferred constr must be big for this

